frc53_nft/src/types.rs

```diff
@@ -159,3 +162,48 @@ pub struct RevokeParams {
 pub struct RevokeForAllParams {
     pub operator: Address,
 }
+
+#[derive(Serialize_tuple, Deserialize_tuple, Clone, Debug)]
+pub struct ListTokensParams {
+    pub cursor: Option<AmtCursor>,
```

Comment:
For the purposes of an API decoupled from implementation, I think we want all the cursor parameters to be `[u8]`, from/into which you can then de/serialise an AmtCursor explicitly. This should probably remain `[u8]` at the library level, but be translated for the state API.

Reply:
I disagree here given that this particular cursor format _is_ inherently tied to this specific reference implementation/library. The library method could take `[u8]` as a parameter but it's going to break anyway if anything other than a serialised `AmtCursor` is passed in so the typesafety at library level does add some value. The API as specified by the FRC dictates actor-actor boundaries, against which `ListTokensParams` is compliant when serialised (if my understanding of the serialisation logic is correct).


If your motivation is more about communicating that the `cursor` can be implementation-specific, then I'd rather add some documentation/examples demonstrating alternate or augmented cursor mechanics.

Reply:
Firstly context: I'm interpreting these types as the ones that should be used by actors implementing tokens to define their binary API. Otherwise I would not expect derived serialization. The built-in datacap actor uses those types already defined here.

This is not compliant with the FRC (which specified IPLD-CBOR tuple encoding). The FRC specifies something of the form
```
[
 h'1234abcd',
 10000
]
```
This will produce instead
```
[
 [
  42(h'5678efgh'),
  100
 ],
 10000
]
```

Where the bytes in `h'1234abcd'` would deserialize to `[42(h'5678efgh'),100]`, and `42(...)` is a CID. (Use cbor.me to inspect those snippets.)

I understand that it will be possible to pass ListTokensParams that deserialize correctly but contain an invalid token. That's fine – there are other ways to get an invalid token too. The library can deserialize the opaque token and check its structure.

Reply:
Made the changes https://github.com/helix-onchain/filecoin/pull/198/commits/85e812eaed6b83dd1dac5d2cd2108946d0fd38f6